python/hail/expr/expressions/base_expression.py

```diff
@@ -365,18 +365,26 @@ def _promote_scalar(self, typ):
             return hail.float64(self)
 
     def _promote_numeric(self, typ):
-        if isinstance(typ, tarray):
-            if isinstance(self.dtype, tarray):
-                return hail.map(lambda x: x._promote_scalar(typ.element_type), self)
-            else:
-                return self._promote_scalar(typ.element_type)
-        elif isinstance(self, expressions.Aggregable):
-            return self._map(lambda x: x._promote_scalar(typ))
+        coercer = expressions.coercer_from_dtype(typ)
+        if isinstance(typ, tarray) and not isinstance(self.dtype, tarray):
+            return coercer.ec.coerce(self)
         else:
-            return self._promote_scalar(typ)
+            return coercer.coerce(self)
 
     def _bin_op_numeric_unify_types(self, name, other):
-        t = unify_types(self.dtype._scalar_type(), other.dtype._scalar_type())
+        def numeric_proxy(t):
```

Comment:
bad name, happy to hear suggestions. This logic is a bug grimy but I don't really see a more elegant way to replace bool with int32. This change means that we can multiply booleans, for instance.